src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.SnapshotManager.cs

```diff
@@ -0,0 +1,202 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed partial class NullableWalker
+    {
+        internal sealed class SnapshotManager
+        {
+            /// <summary>
+            /// The int key corresponds to <see cref="Snapshot.GlobalStateIndex"/>.
+            /// </summary>
+            private readonly ImmutableArray<SharedWalkerState> _walkerGlobalStates;
```

Comment:
>ImmutableArray<SharedWalkerState> _walkerGlobalStates [](start = 29, length = 53)

Is this array necessary? If `SharedWalkerState` were a `class`, could we use a reference to `SharedWalkerState` in `Snapshot` rather than an index into this array?

Reply:
We could potentially do that, but I'm going to leave this as is. I think making `SharedWalkerState` a class would probably introduce more garbage than it would save.

---
In reply to: [289925758](https://github.com/dotnet/roslyn/pull/35850#discussion_r289925758) [](ancestors = 289925758)

Reply:
There shouldn't be many allocations - roughly one instance of `SharedWalkerState` per variable.

But if that is a concern, the fields of `SharedWalkerState` could be copied into `Snapshot`.

---
In reply to: [289940273](https://github.com/dotnet/roslyn/pull/35850#discussion_r289940273) [](ancestors = 289940273,289925758)